Added dynamic authentication with external command (#3187)
gruntwork-io/terragrunt: docs/_docs/04_reference/cli-options.md

- [run-all](#run-all)

Specify the output directory for the `*-all` commands to store plans in JSON format. Useful to read plans programmatically.

### terragrunt-auth-provider-cmd

**CLI Arg**: `--terragrunt-auth-provider-cmd`<br/>
**Environment Variable**: `TERRAGRUNT_AUTH_PROVIDER_CMD`<br/>
**Requires an argument**: `--terragrunt-auth-provider-cmd /path/to/executable/file`<br/>

> The path to the executable file to obtain authentication credentials. If specified, Terragrunt runs this command for every working directory. The output must be the following JSON string

Reviewer: ```suggestion
The path to the executable file to obtain authentication credentials. If specified, Terragrunt runs this command for every working directory.

The output must be valid JSON of the following schema:
```

Then show that it's a map of string to string, and include some keys that are not related to AWS.

Author: Updated, thanks!

- [run-all](#run-all)

Specify the output directory for the `*-all` commands to store plans in JSON format. Useful to read plans programmatically.

### terragrunt-auth-provider-cmd

**CLI Arg**: `--terragrunt-auth-provider-cmd`<br/>
**Environment Variable**: `TERRAGRUNT_AUTH_PROVIDER_CMD`<br/>

> **Requires an argument**: `--terragrunt-auth-provider-cmd /path/to/executable/file`<br/>

Reviewer: ```suggestion
**Requires an argument**: `--terragrunt-auth-provider-cmd "command [arguments]"`<br/>
```

Author: Much better👍 Thanks.

> The path to the executable file to obtain authentication credentials. If specified, Terragrunt runs this command for every working directory.

Reviewer: ```suggestion
The command and arguments used to obtain authentication credentials dynamically. If specified, Terragrunt runs this command for every working directory before running the underlying IAC for a `terragrunt.hcl` file.
```

Author: Thanks!

The output must be valid JSON of the following schema:

```json

> {

Reviewer 2: Maybe make sense to add in example other non `awsCredentials` variables?

Author: Agree, added. Thanks.

  "awsCredentials": {
    "ACCESS_KEY_ID": "",
    "SECRET_ACCESS_KEY": "",
    "SESSION_TOKEN": ""

>   }

Author: Improved the example.

}

> ```

Author: Excellent! I envy your ability to write good documentation. Thanks!

The output must be valid JSON of the following schema:

```json
{
  "awsCredentials": {
    "ACCESS_KEY_ID": "",
    "SECRET_ACCESS_KEY": "",
    "SESSION_TOKEN": ""
  },
  "envs": {
    "ANY_KEY": ""
  }
}
```

This allows Terragrunt to acquire different credentials at runtime without changing any terragrunt.hcl configuration. You can use this flag to set arbitrary credentials for continuous integration, authentication with providers other than AWS and more.

As long as the standard output of the command passed to terragrunt-auth-provider-cmd results in JSON matching the schema above, corresponding environment variables will be set before Terragrunt begins IAC execution for a terragrunt.hcl file.

The simplest approach to leverage this flag is to write a script that fetches desired credentials, and emits them to STDOUT in the JSON format listed above:

```bash
#!/usr/bin/env bash

echo -n '{"envs": {"KEY": "a secret"}}'
```

You can use any technology you'd like, however, as long as Terragrunt can execute it. The expected pattern for using this flag is to populate the values dynamically using a secret store, etc.

> Note that more specific configurations (e.g. awsCredentials) take precedence over less specific configurations (e.g. envs).

Reviewer: ```suggestion
Note that more specific configurations (e.g. `awsCredentials`) take precedence over less specific configurations (e.g. `envs`).
```

NIT

Author: Ah, Sorry for my inattention. Fixed everywhere.

> If you would like to set credentials for AWS with this method, you are encouraged to use awsCredentials instead of envs, as these keys will be validated to conform to the officially supported environment variables expected by the AWS SDK.

Reviewer: ```suggestion
If you would like to set credentials for AWS with this method, you are encouraged to use `awsCredentials` instead of `envs`, as these keys will be validated to conform to the officially supported environment variables expected by the AWS SDK.
```

NIT

> Other credential configurations will be supported in the future, but until then, if your provider authenticates via environment variables, you can use the envs field to fetch credentials dynamically from a secret store, etc before Terragrunt executes any IAC.

Reviewer: ```suggestion
Other credential configurations will be supported in the future, but until then, if your provider authenticates via environment variables, you can use the `envs` field to fetch credentials dynamically from a secret store, etc before Terragrunt executes any IAC.
```
